### PR TITLE
Track state in summingbird-online as an Iterator rather than a Seq.

### PR DESCRIPTION
**Pull request by pankajroark**
Fix for #689 . This should avoid n^2 compute complexity when summing single element lists of Storm tuples.

### CODE REVIEW

**Comment by johnynek** [comment]
could we use `Stream` instead? the mutability of an `Iterator` makes it pretty costly to verify that it is not buggy with a code review. I really hate to have a mutable object on an API unless the performance is really necessitating it?

**Comment by pankajroark** [comment]
Good idea. Changed to use stream instead of iterator.

**Comment by pankajroark** [comment]
tbh Iterator is 4x faster than stream in this microbenchmark but both are way faster than List, so stream seems fine:
    val s = (0 to 10000).toList

  @Benchmark
  def listConcat(): List[Int] = {
    s ++ List(0)
  }

  @Benchmark
  def streamConcat(): Stream[Int] = {
    s.toStream ++ Stream(0)
  }

  @Benchmark
  def iterConcat(): Iterator[Int] = {
    s.toIterator ++ Iterator.single(0)
  }

Results:
[info] ToBenchmark.iterConcat    thrpt       4  85182247.256 ± 5840222.962  ops/s
[info] ToBenchmark.listConcat    thrpt        4         9937.918 ±   26498.037  ops/s
[info] ToBenchmark.streamConcat  thrpt   4  22019587.193 ± 7199199.008  ops/s

**Comment by johnynek** [comment]
I would bet Batched will be as fast or faster than Iterator.

**Comment by pankajroark** [comment]
I tried Batched but the code started getting complicated because of https://github.com/twitter/summingbird/blob/develop/summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala#L46

We need to be able to pass an empty Batched here. Batched itself doesn't have a zero value it relies on the contained type. So we either have to use Option[Batched] which wouldn't be good for performance, always having to wrap/unwrap. Or we have to have the monoid of S available in this abstract class, which would require further code changes.

**Comment by pankajroark** [comment]
Just for comparison iterator and Batched are indeed comparable in perf. I'd really like to use Batched if we could find a simple way or iterator otherwise. Combining input state is in the hot path.:
  val s = (0 to 10000).toList
  val bs = Batched.items(s).get
  val is = s.toIterator
  val ss = s.toStream

  @Benchmark
  def listConcat(): List[Int] = {
    s ++ List(0)
  }

  @Benchmark
  def streamConcat(): Stream[Int] = {
    ss ++ Stream(0)
  }

  @Benchmark
  def iterConcat(): Iterator[Int] = {
    is ++ Iterator.single(0)
  }

  @Benchmark
  def batchedConcat(): Batched[Int] = {
    bs.combine(Batched(0))
  }

[info] ToBenchmark.batchedConcat  thrpt    4  126300963.122 ± 46804561.489  ops/s
[info] ToBenchmark.iterConcat     thrpt         4  121149476.337 ± 46339829.598  ops/s
[info] ToBenchmark.listConcat     thrpt         4            9380.640 ±    23140.126  ops/s
[info] ToBenchmark.streamConcat   thrpt    4   24118157.945 ±  7172767.691  ops/s

**Comment by pankajroark** [comment]
Any suggestions on the next steps here. I'm ok with Stream, it's still much better than List in this case.

**Comment by johnynek** [comment]
I have not investigated why the tests are red.

If we can make them green with something faster. I'd be happy to do that.

I don't think using Batched would be too much work. Or we could copy this code or add this dependency:

https://github.com/non/chain/blob/master/src/main/scala/chain/Chain.scala

It is a single file library that is the more general version of Batched (it has empty), it was written by @non who also wrote Batched.

Do we really need an empty Batched? I would imagine that a `Monoid[Option[Batched[T]]]` would be almost as fast (still faster than Stream).

It is up to you. I think killing the O(N^2) is most important. Losing a constant factor of 4 is probably not a huge deal if you don't want to work on this other stuff.

Not using a mutable data structure is pretty important to me since this code has been worked on by many people now, and it is much easier to make a mistake with mutable APIs.

**Comment by pankajroark** [comment]
Thanks let me try chained. I agree, constant factor of 4 is insignificant
compared to n^2. Tests are failing the Mima check.

**Comment by johnynek** [comment]
Can you add the exclusions to so we can keep the tests green? It prints
what you need to add to the build to exclude those methods from erroring.

**Comment by pankajroark** [comment]
Sure I'll take a look later today.

**Comment by pankajroark** [comment]
Chain seems good:
[info] ToBenchmark.batchedConcat  thrpt    4  133738359.414 ±  3281958.077  ops/s
[info] ToBenchmark.chainConcat    thrpt    4      90914620.450 ± 19299667.614  ops/s
[info] ToBenchmark.iterConcat     thrpt    4       122448980.991 ± 38379171.178  ops/s
[info] ToBenchmark.streamConcat   thrpt    4     26017386.004 ±  6404231.437  ops/s

I've updated the review with now using chain.

**Comment by pankajroark** [comment]
Tests pass now, just waiting for shipit to merge.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/BaseBolt.scala [hunk 1]
```diff
@@ -149,12 +150,13 @@ case class BaseBolt[I, O](jobID: JobId,
     }
   }
 
-  private def finish(inputs: Stream[InputState[Tuple]], results: TraversableOnce[O]) {
+  private def finish(inputs: Chain[InputState[Tuple]], results: TraversableOnce[O]) {
+    val tuples = inputs.iterator.map(_.state).toList
```
Comment (johnynek) [inline summingbird-storm/src/main/scala/com/twitter/summingbird/storm/BaseBolt.scala]: shouldn't we move this line to 158 (only materialize the `List` if we have dependants and we anchor?

**Review by johnynek — CHANGES_REQUESTED** [review]

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/BaseBolt.scala (hunk 1 above)
Comment (pankajroark) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/BaseBolt.scala]: My reasoning was that we need to iterate through the chain anyway to get the size for the log statement at the end of function but I agree, materialization of list should be avoided as well in those cases. Proposed a fix.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/BaseBolt.scala [hunk 2]
```diff
@@ -151,11 +151,13 @@ case class BaseBolt[I, O](jobID: JobId,
...
 
   private def finish(inputs: Chain[InputState[Tuple]], results: TraversableOnce[O]) {
-    val tuples = inputs.iterator.map(_.state).toList
+    var numTuples = Option.empty[Int]
     var emitCount = 0
     if (hasDependants) {
       if (anchorTuples.anchor) {
         results.foreach { result =>
+          val tuples = inputs.iterator.map(_.state).toList
+          numTuples = Some(tuples.size)
```
Comment (johnynek) [inline summingbird-storm/src/main/scala/com/twitter/summingbird/storm/BaseBolt.scala]: list `.size` is `O(N)` not `O(1)`. So if we are going to do this, why use the var and not just use use `inputs.iterator.size`.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/BaseBolt.scala [hunk 3]
```diff
@@ -151,11 +151,13 @@ case class BaseBolt[I, O](jobID: JobId,
   }
 
   private def finish(inputs: Chain[InputState[Tuple]], results: TraversableOnce[O]) {
-    val tuples = inputs.iterator.map(_.state).toList
+    var numTuples = Option.empty[Int]
     var emitCount = 0
     if (hasDependants) {
       if (anchorTuples.anchor) {
         results.foreach { result =>
+          val tuples = inputs.iterator.map(_.state).toList
```
Comment (johnynek) [inline summingbird-storm/src/main/scala/com/twitter/summingbird/storm/BaseBolt.scala]: don't we want this above the foreach? We don't want to recompute `tuples` for each result. We just want it once and then reuse for all results, no?

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/BaseBolt.scala (hunk 2 above)
Comment (pankajroark) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/BaseBolt.scala]: Good point. We should be able to calculate the size and list in one pass though. Let me try.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/BaseBolt.scala (hunk 3 above)
Comment (pankajroark) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/BaseBolt.scala]: Yeah, good idea. I believe this was how it was in original code too so I didn't notice. Let me try to fix this.

**Comment by johnynek** [comment]
:+1:

**Comment by ttim** [comment]
What is a reason to expose `Chain` on a level of `OperationContainer` ? I'm not strictly against that but `Traversable` or even `TraversableOnce` seems more appropriate for me. 

To be more precise - `AsyncSummer` instances should be over `Chain` to avoid N^2 complexity while everything else should be over `Traversable`/`TraversableOnce`. What do you think?

**Comment by johnynek** [comment]
I think making a follow up PR that minimized the scope of visibility of Chain (maybe even walking back some of the mima exclusions) would be fine. But if we have to copy to do it, I would not.

For instance, you can go: `Iterable[T] => Chain[T]` but not `Iterator[T] => Chain[T]` without a copy.

So, maybe we could use `Iterable[T]` in some cases, and internally use a `Chain[T]` to do fast concat.

That said, these are fairly "private" classes in that 99% of summingbird users would never use them. In fact, likely only storm/heron platform would use them.

**Comment by ttim** [comment]
I agree.

Regarding to copying - we don't need `Iterator[T] => Chain[T]` transformation because `OperationContainer#execute` accepts single `state` element. But we need `Chain[T] => Iterable[T]` transformation which is the same (in terms of copying) as `Chain[T] => Iterator[T]`.

**Comment by pankajroark** [comment]
I agree exposing chain at OperationContainer is not ideal and also that OperationContainer is used only in storm/heron platform right now, so I feel it's better to keep it simple, as is, for now. Let me create an issue to capture this though.

**Pull request merged**